### ISSUE TITLE
Pillow and Tkinter produce Segmentation fault

### ISSUE THREAD

**Issue by eine** [msys2/MINGW-packages#7136]
I'm trying to test Tkinter and Pillow on MINGW64 but I get a Segmentation fault.

The following minimal example works and shows an empty window:

```py
import tkinter
top = tkinter.Tk()
# Code to add widgets will go here...
top.mainloop()
```

This one produces a segfault:

```py
# tkinter_pillow.py

from tkinter import Tk, Frame, BOTH
from PIL import Image, ImageTk

class Example(Frame):

    def __init__(self):
        super().__init__()

        self.master.title("High Tatras")
        self.pack(fill=BOTH, expand=1)

        self.img = Image.open("test.jpg")
        self.tatras = ImageTk.PhotoImage(self.img)

root = Tk()
ex = Example()
root.mainloop()
```

It should show nothing, because it's incomplete, but it should not crash.

I tried the following example too: https://raw.githubusercontent.com/python-pillow/Pillow/3.0.x/Scripts/viewer.py. It also produces a Segmentation fault.

# How to reproduce

I tried the following in a fresh https://github.com/msys2/msys2-installer/releases/download/nightly-x86_64/msys2-base-x86_64-latest.tar.xz:

```sh
# pacman -Syu
...
# pacman -S mingw-w64-x86_64-python-pip
...
# pacman -S mingw-w64-x86_64-python-pillow
...
# python3 tkinter_pillow.py
Segmentation fault

# python3 viewer.py test.jpg
Segmentation fault

# pacman -S mingw-w64-x86_64-python-pytest
...

#  git clone https://github.com/python-pillow/Pillow
...

# cd Pillow

# PYTHONPATH=$(pwd) python3 -m pytest Tests/test_imagetk.py
============================= test session starts =============================
platform win32 -- Python 3.8.6, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
pytest_report_header failed: pilinfo() got an unexpected keyword argument 'supported_formats'
rootdir: C:/Users/eine/Downloads/tkinter/Pillow, inifile: setup.cfg
collected 4 items

Tests/test_imagetk.py .Windows fatal exception: access violation

Current thread 0x00007714 (most recent call first):
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/PIL/ImageTk.py", line 203 in paste
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/PIL/ImageTk.py", line 121 in __init__
  File "C:/Users/eine/Downloads/tkinter/Pillow/Tests/test_imagetk.py", line 61 in test_photoimage
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/python.py", line 184 in pytest_pyfunc_call
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/python.py", line 1479 in runtest
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/runner.py", line 135 in pytest_runtest_call
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/runner.py", line 217 in <lambda>
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/runner.py", line 244 in from_call
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/runner.py", line 216 in call_runtest_hook
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/runner.py", line 186 in call_and_report
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/runner.py", line 100 in runtestprotocol
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/runner.py", line 85 in pytest_runtest_protocol
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/main.py", line 272 in pytest_runtestloop
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/main.py", line 247 in _main
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/main.py", line 191 in wrap_session
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/main.py", line 240 in pytest_cmdline_main
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/_pytest/config/__init__.py", line 124 in main
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/site-packages/pytest/__main__.py", line 7 in <module>
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/runpy.py", line 87 in _run_code
  File "C:/Users/eine/Downloads/msys64/mingw64/lib/python3.8/runpy.py", line 194 in _run_module_as_main
Segmentation fault
```

**Comment by jeremyd2019** [comment]
It appears that this upstream patch fixes this crash:
https://github.com/python-pillow/Pillow/commit/3273a1af4c4e03559528301a4e4681a453709c26

**Comment by nulano** [comment]
The `mingw-w64-x86_64-python-pillow` package is an old version of Pillow, 6.2.1 IIRC. The fix was released as part of Pillow 8.0.0. You can use `python -m pip install Pillow` instead of `pacman -S` to install the latest version. Note that the patches for Pillow in this repository are no longer needed as the issues have been fixed upstream.

See instructions for installing from source here: https://pillow.readthedocs.io/en/stable/installation.html#building-on-windows-using-msys2-mingw

**Comment by eine** [comment]
I can confirm that installing all the dependencies listed in https://pillow.readthedocs.io/en/stable/installation.html#building-on-windows-using-msys2-mingw allows installing Pillow through pip. And the latest version installed through pip does not crash. Thanks @nulano!

@lazka, shall `mingw-w64-x86_64-python-pillow` be removed?

**Comment by jeremyd2019** [comment]
Or updated, or patched against this issue if there's some good reason for staying with the old version...

**Comment by lazka** [comment]
> @lazka, shall `mingw-w64-x86_64-python-pillow` be removed?

It's a dependency of various other packages, so we can't do that.

8.0 has been released quite recently and it's not clear if all other packages work with it, so a safe fix would be to update to 7.x and backport that patch as a first step.

**Comment by eine** [comment]
> Note that the patches for Pillow in this repository are no longer needed as the issues have been fixed upstream.

@nulano, are the patches fixed in v8.0.0 only or in v7.2.0 too? I see no references to MSYS2 in the changelog.

**Comment by nulano** [comment]
The fix is in 8.0.0 only.

> if there's some good reason for staying with the old version...

One possible reason is that 6.2.x is the last series of Pillow with Python 2.7 support, but I would expect that to only be a problem for the Python 2 version of the package, not the Python 3 version.

**Comment by eine** [comment]
> The fix is in 8.0.0 only.

I mean the other patches (https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-python-pillow), not the one about the segfault.

**Comment by nulano** [comment]
Ah, sorry. `001-freeze-support.patch` is invalid since 7.0.0 when Python 2.7 was dropped, I'm not sure if it is still needed. The other patches were fixed in 7.2.0.

While upgrading you might also run into an issue with LibTIFF, which was fixed in https://github.com/python-pillow/Pillow/pull/4890/commits/3cabcf242f472d0aa99ce03f6164787d902091a7 released in 8.0.0.

**Comment by lazka** [comment]
I've created #7139
